Document ts project references setup (#78586)
elastic/kibana: docs/development/typescript.md

# Typescript
Although this is not a requirement, we encourage if all new code is developed in [Typescript](https://www.typescriptlang.org/).

- [Project references](#project-references)
  - [Caveats](#caveats)
  - [Prerequisitions](#prerequisitions)
  - [Implementation](#implementation)

## Project references

> Kibana has crossed the 1.5m LoC mark. The current situation creates some scaling problems when the default out-of-the-box setup stops working. As a result, developers suffer from slow project compilation and IDE unresponsiveness. As a part of [Developer Experience project](https://github.com/elastic/kibana/projects/63), we are migrating our tooling to use built-in TypeScript features addressing the scaling problems - [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) & [incremental buils](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#faster-subsequent-builds-with-the---incremental-flag). In a nutshell - instead of compiling the whole Kibana codebase at once, this setup enforces splitting the code base into independent projects that form a directed acyclic graph (DAG) that allows the TypeScript compiler(`tsc`) to apply several advanced optimizations:

Reviewer: ```suggestion
Kibana has crossed the 1.5m LoC mark. The current situation creates some scaling problems when the default out-of-the-box setup stops working. As a result, developers suffer from slow project compilation and IDE unresponsiveness. As a part of [Developer Experience project](https://github.com/elastic/kibana/projects/63), we are migrating our tooling to use built-in TypeScript features addressing the scaling problems - [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) & [incremental builds](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#faster-subsequent-builds-with-the---incremental-flag).

In a nutshell - instead of compiling the whole Kibana codebase at once, this setup enforces splitting the code base into independent projects that form a directed acyclic graph (DAG). This allows the TypeScript compiler (`tsc`) to apply several advanced optimizations:
```

Reviewer: Diff isn't making this obvious, but I fixed a typo or two in there as well ^

> - every project emits *public* interfaces in the form of `d.ts` type declaration
> - created`d.ts` type declarations are used whenever a referenced project is imported in a depending project
> - it makes it possible to determine which project needs rebuilding when the source code has changed to use a more aggressive caching strategy.
> More details are available in the [official docs](https://www.typescriptlang.org/docs/handbook/project-references.html)

Reviewer: ```suggestion
- Every project emits *public* interfaces in the form of `d.ts` type declarations generated by the TypeScript compiler
- These generated `d.ts` type declarations are used whenever a referenced project is imported in a depending project
- This makes it possible to determine which project needs rebuilding when the source code has changed to use a more aggressive caching strategy.
More details are available in the [official docs](https://www.typescriptlang.org/docs/handbook/project-references.html)
```

### Caveats
This architecture imposes several limitations to which we must comply:

> - projects cannot have circular dependencies. Even though the Kibana platform doesn't support circular dependencies between Kibana plugins, TypeScript does allow circular type imports between files. So in theory, you might face a problem when migrating to the TS project references so you will have to remove this circular dependency.
> - a project must emit its type declaration. It's not always possible to generate a type declaration if the compiler cannot reference a type. There are two basic cases:

Reviewer: ```suggestion
- Projects cannot have circular dependencies. Even though the Kibana platform doesn't support circular dependencies between Kibana plugins, TypeScript (and ES6 modules) does allow circular imports between files. So in theory, you may face a problem when migrating to the TS project references and you will have to resolve this circular dependency.
- A project must emit its type declaration. It's not always possible to generate a type declaration if the compiler cannot infer a type. There are two basic cases:
```

1. Your plugin exports a type inferring an internal type declared in Kibana codebase. In this case, you'll have to either export an internal type or to declare an exported type explicitly.
2. Your plugin exports something inferring a type from a 3rd party library that doesn't export this type. To fix the problem, you have to declare the exported type manually.

### Prerequisitions

> Since `tsc` doesn't support circular references, the migration order does matter. You can migrate your plugin only when all the plugin dependencies already have migrated. It creates the situation when low-level plugins (such as `data` or `kibana_react`) have to migrate first.

Reviewer 2: It would be useful to add a script showing a plugin dependency tree. I didn't manage to find an issue - I can create one. Should we prioritize it? Manual migration tracking is not cool.

Reviewer: Could probably be solved by the same tool as https://github.com/elastic/kibana/issues/78162 ?

Reviewer 2: Yes, it would help to understand all the existing deps between plugins. But:
1. It's no clear when https://github.com/elastic/kibana/issues/78162 is taken in work by the @elastic/kibana-operations 
2. A simple script visualizing a plugin tree from the plugin discovery would do the trick as well. As a bonus, we could highlight all the projects already migrated to TS project refs.

The main problem is the plugin discovery is tightly coupled with the platform runtime. We might have to either extract it in or borrow `@kbn/optimizer` implementation https://github.com/elastic/kibana/blob/7c6a40ea34872a1c08ac5c1e01c19c99e1742030/packages/kbn-optimizer/src/optimizer/optimizer_config.ts#L213

> Since `tsc` doesn't support circular references, the migration order does matter. You can migrate your plugin only when all the plugin dependencies already have migrated. It creates the situation when low-level plugins (such as `data` or `kibana_react`) have to migrate first.

Reviewer: ```suggestion
Since `tsc` doesn't support circular project references, the migration order does matter. You can migrate your plugin only when all the plugin dependencies already have migrated. It creates a situation where commonly used plugins (such as `data` or `kibana_react`) have to migrate first.
```

> ### Implementation

Reviewer: nit: do you mind capitalizing all of these bullet points

- make sure all the plugins listed as dependencies in `kibana.json` file have migrated to TS project references.
- add `tsconfig.json` in the root folder of your plugin.
```json
{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": {
    "composite": true,
    "outDir": "./target/types",
    "emitDeclarationOnly": true,
    "declaration": true,
    "declarationMap": true
  },
  "include": [
    // add all the folders containg files to be compiled
  ],
  "references": [
    { "path": "../../core/tsconfig.json" },
    // add references to other TypeScript projects your plugin dependes on
  ]
}
```
If your plugin imports a file not listed in `include`, the build will fail with the next message `File ‘…’ is not listed within the file list of project …’. Projects must list all files or use an 'include' pattern.`
- build you plugin `./node_modules/.bin/tsc -b src/plugins/my_plugin`. Fix errors if `tsc` cannot generate type declarations for your project.

> - make sure the `target/types` folder doesn’t contain files not belonging to your project. Otherwise, it means your plugin includes files from another project by mistake.

Reviewer: Is there an easy way to detect this?

Reviewer 2: We can write a script to make sure all files in outDir covered by patterns in `includes`. However, if code from another plugin is compiled, it means that this plugin has been listed in `include`. I'm leaning towards removing this note from the doc. WDYT?

Reviewer: If it's not expected to be a common situation, then let's either remove it or move it to a "troubleshooting" section.

- A project must emit its type declaration. It's not always possible to generate a type declaration if the compiler cannot infer a type. There are two basic cases:
1. Your plugin exports a type inferring an internal type declared in Kibana codebase. In this case, you'll have to either export an internal type or to declare an exported type explicitly.
2. Your plugin exports something inferring a type from a 3rd party library that doesn't export this type. To fix the problem, you have to declare the exported type manually.

### Prerequisitions
Since `tsc` doesn't support circular project references, the migration order does matter. You can migrate your plugin only when all the plugin dependencies already have migrated. It creates a situation where commonly used plugins (such as `data` or `kibana_react`) have to migrate first.

### Implementation

> - make sure all the plugins listed as dependencies in `kibana.json` file have migrated to TS project references.

Reviewer 3: nit: I think a numbered list makes more sense here since you will generally start at the top and work down

Reviewer 2: I had to revert it in https://github.com/elastic/kibana/pull/78586/commits/6efc968f8595599d5560d7580f9de812137b7ab9 to the unordered list. asciidoctor fails validation with
```
18:06:10 INFO:build_docs:asciidoctor: WARNING: developer/best-practices/typescript.asciidoc: line 35: list item index: expected 2, got 1
18:06:10 INFO:build_docs:asciidoctor: WARNING: developer/best-practices/typescript.asciidoc: line 59: list item index: expected 2, got 1
18:06:10 INFO:build_docs:asciidoctor: WARNING: developer/best-practices/typescript.asciidoc: line 60: list item index: expected 3, got 1
18:06:10 INFO:build_docs:asciidoctor: WARNING: developer/best-practices/typescript.asciidoc: line 61: list item index: expected 4, got 1
18:06:10 INFO:build_docs:asciidoctor: WARNING: developer/best-practices/typescript.asciidoc: line 62: list item index: expected 5, got 1
```
I tried `1., 2., etc.` style as well, but the numeration started from `1` after `source` block. 😞

- add `tsconfig.json` in the root folder of your plugin.
```json
{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": {
    "composite": true,
    "outDir": "./target/types",
    "emitDeclarationOnly": true,
    "declaration": true,
    "declarationMap": true
  },
  "include": [
    // add all the folders containg files to be compiled
  ],
  "references": [
    { "path": "../../core/tsconfig.json" },
    // add references to other TypeScript projects your plugin dependes on
  ]
}
```
If your plugin imports a file not listed in `include`, the build will fail with the next message `File ‘…’ is not listed within the file list of project …’. Projects must list all files or use an 'include' pattern.`
- build you plugin `./node_modules/.bin/tsc -b src/plugins/my_plugin`. Fix errors if `tsc` cannot generate type declarations for your project.
- make sure the `target/types` folder doesn’t contain files not belonging to your project. Otherwise, it means your plugin includes files from another project by mistake.

> - add your project reference to `include` property of `tsconfig.refs.json`

Reviewer 3: Shouldn't this be?
```suggestion
- add your project reference to the `references` property of `tsconfig.refs.json`
```

> - add your plugin to `include` property and plugin folder to `exclude` property of the `tsconfig.json` it used to belong to (for example, for `src/plugins/**` it's `tsconfig.json`; for `x-pack/plugins/**` it’s `x-pack/tsconfig.json`).

Reviewer 3: Is this also the `references` property instead of `include` property?

Reviewer 2: right
